contracts/bid/src/tree.rs

```diff
@@ -0,0 +1,242 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use super::leaf::BidLeaf;
+use crate::contract_constants::BID_TREE_DEPTH;
+use canonical::{Canon, Store};
+use canonical_derive::Canon;
+use dusk_poseidon::tree::{
+    PoseidonBranch, PoseidonMaxAnnotation, PoseidonTree, PoseidonTreeIterator,
+};
+use microkelvin::{BranchMut, Nth};
+use nstack::NStack;
+
+/// Append-only tree structure which wraps over `PoseidonTree<BidLeaf>`
+/// responsible of the storage and the branch generation capabilities
+/// of the Bid contract.
+#[derive(Default, Debug, Clone, Canon)]
+pub struct BidTree<S>
+where
+    S: Store,
+{
+    tree: PoseidonTree<BidLeaf, PoseidonMaxAnnotation, S, BID_TREE_DEPTH>,
+}
+
+impl<S> BidTree<S>
+where
+    S: Store,
+{
+    /// Constructor for the Tree structure.
+    pub fn new() -> Self {
+        Self {
+            tree: PoseidonTree::new(),
+        }
+    }
+
+    /// Reference to the internal poseidon tree
+    ///
+    /// We don't have a mutable reference available because all its mutation
+    /// should be protected by encapsulation
+    fn inner_mut(
+        &mut self,
+    ) -> &mut PoseidonTree<BidLeaf, PoseidonMaxAnnotation, S, BID_TREE_DEPTH>
+    {
+        &mut self.tree
+    }
+
+    /// Get a bid from a provided index
+    pub fn get(&self, idx: u64) -> Option<BidLeaf> {
+        self.tree
+            .get(idx as usize)
+            .expect("Couldn't traverse the tree.")
```

Comment:
Does it really make sense to unwrap/expect outside the explicit contract functions?

Reply:
I did it so that I don't need to add more arms to match statements. Anyway this will always be called in the contract functions.